[PF-529] Add functionality to update a workspace.

tnadler opened the pull request:
- Define new API for UpdateWorkspace.
- Add displayName and description fields as needed.
- Implement UpdateWorkspace in the controller and the service.

zloery commented on workspace-manager-clienttests/src/main/resources/configs/integration/WorkspaceLifecycle.json:
```diff
@@ -1,12 +1,12 @@
 {
-  "name": "CreateGetDeleteWorkspace",
```
A couple extra changes to make with this:

- This file is the integration test case, but we have a similar test for performance with different parameters at workspace-manager-clienttests/src/main/resources/configs/perf/CreateGetDeleteWorkspace.json. could you update that as well?
- This file (and the perf test) is referenced by a test suite, please update the name there as well (workspace-manager-clienttests/src/main/resources/suites/BasicIntegration.json and BasicPerf.json)

tnadler replied:
Done (I think)

zloery commented on src/main/java/bio/terra/workspace/app/controller/WorkspaceApiController.java:
```diff
@@ -161,19 +166,22 @@ private ApiWorkspaceDescription buildWorkspaceDescription(Workspace workspace) {
     Workspace workspace = workspaceService.getWorkspace(id, userReq);
-    ApiGcpContext gcpContext =
-        workspace.getGcpCloudContext().map(GcpCloudContext::toApi).orElse(null);
+    ApiWorkspaceDescription desc = buildWorkspaceDescription(workspace);
```
Whoops - not sure how I missed this earlier. good catch!

zloery commented on src/main/java/bio/terra/workspace/app/controller/WorkspaceApiController.java:
```diff
@@ -161,19 +166,22 @@ private ApiWorkspaceDescription buildWorkspaceDescription(Workspace workspace) {
-    logger.info("Got workspace {} for {}", desc, userReq.getEmail());
+  @Override
+  public ResponseEntity<ApiWorkspaceDescription> updateWorkspace(
+      @PathVariable("workspaceId") UUID id, @RequestBody ApiUpdateWorkspaceRequestBody body) {
```
nit: first parameter should be `workspaceId` to match the underlying interface. 
I think the parameter name was changed without updating existing controller methods, so there are probably inconsistencies elsewhere in this file

ddietterich replied:
Yes, I think I did the openApi change, but didn't change the code.

tnadler replied:
Done

zloery reviewed (commented):
Looks good, a few extra changes needed for the client tests.
The new connected tests are also failing, though I didn't dig into causes to see how real those were

ddietterich commented on src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java:
```diff
@@ -126,6 +127,22 @@ public Workspace getWorkspace(UUID id, AuthenticatedUserRequest userReq) {
+      UUID workspaceId,
+      @Nullable String name,
+      @Nullable String description) {
+    validateWorkspaceAndAction(userReq, workspaceId, SamConstants.SAM_WORKSPACE_WRITE_ACTION);
```
I think we want to get the workspace and report any 404 before we do the validate. This is an extra step, but is nicer to the caller and is the convention.
@zloery check me on this, please

tnadler replied:
I am not sure I understand... This is the function that getWorkspace calls?

ddietterich replied:
Nevermind The validate call does the right thing.

ddietterich commented on src/test/java/bio/terra/workspace/service/workspace/WorkspaceServiceTest.java:
```diff
@@ -209,6 +209,58 @@ void testWithSpendProfile() {
+    // Since name is null, leave it alone. Description should be updated.
+    assertEquals(name, secondUpdatedWorkspace.getDisplayName());
+    assertEquals(otherDescription, secondUpdatedWorkspace.getDescription());
+
```
Maybe test the both passed as null should result in an error.

tnadler replied:
Done

ddietterich commented on workspace-manager-clienttests/src/main/java/scripts/testscripts/WorkspaceLifecycle.java:
```diff
@@ -31,6 +32,12 @@ public void doUserJourney(TestUserSpecification testUser, WorkspaceApi workspace
+    UpdateWorkspaceRequestBody updateBody = new UpdateWorkspaceRequestBody().displayName("name").description("desc");
+    WorkspaceDescription updatedDescription = workspaceApi.updateWorkspace(updateBody, workspaceId);
+    ClientTestUtils.assertHttpSuccess(workspaceApi, "PATCH workspace");
+    assertThat(updatedDescription.getDisplayName(), equalTo("name"));
```
Maybe make the strings static?

tnadler replied:
Done

ddietterich reviewed (commented):
Mostly nits; one change to check existence before access in the update.

tnadler commented:
Addressed comments and actually needed a bit more logic in the create. Current connected test failure seems spurious. Would love your thoughts on the Optional handling in the create.

ddietterich commented on src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java:
```diff
@@ -78,6 +79,10 @@ public UUID createWorkspace(WorkspaceRequest workspaceRequest, AuthenticatedUser
     createJob.addParameter(
         WorkspaceFlightMapKeys.WORKSPACE_STAGE, workspaceRequest.workspaceStage());
 
+    createJob.addParameter(
```
Instead of passing each element of the request into the flight, we should pass the whole `workspaceRequest` as the request - instead of the `null` that is floating there in the `newJob()` call.

If there are serdes issues, they should be addressed in the WorkspaceRequest object. This simplifies request changes going forward. It can be done as a separate PR. If you choose that, can you make a ticket?

tnadler replied:
Sure. I will file a follow up ticket so I can get this submitted.

ddietterich reviewed (approved):
One comment - probably not for this PR. Up to you.
LGTM
